pkg/ottl/contexts/internal/ctxcontext/context.go

```diff
@@ -0,0 +1,306 @@
+	}
+	return val
+}
+
+func convertGRPCMetadataToMap(md metadata.MD) pcommon.Map {
+	mdMap := pcommon.NewMap()
+	for k, v := range md {
+		sl := mdMap.PutEmptySlice(k)
+		mdSl := convertStringArrToValueSlice(v)
+		mdSl.Slice().CopyTo(sl)
+	}
+	return mdMap
+}
+
+func getIndexableValueFromStringArr[K any](ctx context.Context, tCtx K, keys []ottl.Key[K], strArr []string) (any, error) {
+	val := convertStringArrToValueSlice(strArr)
+	return ctxutil.GetIndexableValue[K](ctx, tCtx, val, keys[1:])
+}
+
+func accessGRPCMetadataKeys[K any]() ottl.StandardGetSetter[K] {
+	return ottl.StandardGetSetter[K]{
+		Getter: func(ctx context.Context, _ K) (any, error) {
+			md, ok := metadata.FromIncomingContext(ctx)
+			if !ok {
+				return pcommon.NewMap(), nil
+			}
+			return convertGRPCMetadataToMap(md), nil
+		},
+		Setter: func(_ context.Context, _ K, _ any) error {
+			return errors.New("cannot set value in context.grpc.metadata")
+		},
+	}
+}
+
+func accessGRPCMetadataKey[K any](keys []ottl.Key[K]) ottl.StandardGetSetter[K] {
+	return ottl.StandardGetSetter[K]{
+		Getter: func(ctx context.Context, tCtx K) (any, error) {
+			if len(keys) == 0 {
+				return nil, errors.New("cannot get map value without keys")
+			}
+			md, ok := metadata.FromIncomingContext(ctx)
+			if !ok {
+				return nil, nil
+			}
+			key, err := ctxutil.GetMapKeyName(ctx, tCtx, keys[0])
+			if err != nil {
+				return nil, err
+			}
+			mdVal := md.Get(*key)
+			if len(mdVal) == 0 {
+				return nil, nil
+			}
+			return getIndexableValueFromStringArr(ctx, tCtx, keys, mdVal)
+		},
+		Setter: func(_ context.Context, _ K, _ any) error {
+			return errors.New("cannot set value in context.grpc.metadata")
+		},
+	}
+}
+
+func accessClientAuth[K any](path ottl.Path[K]) (ottl.GetSetter[K], error) {
+	nextPath := path.Next()
+	if nextPath == nil {
+		return nil, ctxerror.New(path.Name(), path.String(), Name, DocRef)
+	}
+	switch nextPath.Name() {
+	case "attributes":
+		if nextPath.Keys() == nil {
+			return accessClientAuthAttributesKeys[K](), nil
+		}
+		return accessClientAuthAttributesKey[K](nextPath.Keys()), nil
+	default:
+		return nil, ctxerror.New(nextPath.Name(), nextPath.String(), Name, DocRef)
+	}
+}
+
+func getAuthAttributeValue(attr any) (any, error) {
+	switch a := attr.(type) {
+	case string:
+		return a, nil
+	case nil:
+		return nil, nil
+	default:
+		b, err := json.Marshal(attr)
+		if err != nil {
+			return "", err
+		}
+		return string(b), nil
+	}
+}
+
+func convertAuthDataToMap(authData client.AuthData) (pcommon.Map, error) {
+	authMap := pcommon.NewMap()
+	if authData == nil {
+		return authMap, nil
+	}
+	names := authData.GetAttributeNames()
+	for _, name := range names {
+		attrVal := authData.GetAttribute(name)
+		attrStr, err := getAuthAttributeValue(attrVal)
+		if err != nil {
+			return pcommon.NewMap(), err
+		}
+		switch v := attrStr.(type) {
+		case string:
+			authMap.PutStr(name, v)
+		default:
+			authMap.PutEmpty(name)
+		}
+	}
+	return authMap, nil
```

Comment:
```suggestion
	for _, name := range names {
		attrVal := authData.GetAttribute(name)
		authMapAttrVal := authMap.PutEmpty(name)
		switch typedAttrVal := attrVal.(type) {
		case string:
			authMapAttrVal.SetStr(typedAttrVal)
		case []string:
			slice := authMapAttrVal.SetEmptySlice()
			slice.EnsureCapacity(len(typedAttrVal))
			for _, str := range typedAttrVal {
				slice.AppendEmpty().SetStr(str)
			}
		default:
			_ = authMapAttrVal.FromRaw(attrVal)
		}
	}
	return authMap, nil
```

According to the `AuthData.GetAttribute`, it normally returns a `string`, and in a few cases`[]string`.  I got the JSON marshalling fallback idea, but IMO, we shouldn't handle that yet, at least not on the first version. 
If for some reason the data the `GetAttribute` returns is not supported by OTTL/PData, we can keep it empty, and handle those special cases one-by-one as they appear. WDYT?